paasta_tools/autoscaling_lib.py

```diff
@@ -569,104 +589,165 @@ def sort_slaves_to_kill(mesos_state, pool='default'):
-                                                                                            instances_to_kill))
+                                                                                            slave['pid']))
         try:
-            ec2_client.terminate_instances(InstanceIds=instances_to_kill, DryRun=dry_run)
+            ec2_client.terminate_instances(InstanceIds=instance_id, DryRun=dry_run)
         except ClientError as e:
             if e.response['Error'].get('Code') == 'DryRunOperation':
                 pass
             else:
                 raise
 
 
+def get_instance_type_weights(sfr):
+    launch_specifications = sfr['SpotFleetRequestConfig']['LaunchSpecifications']
+    return {ls['InstanceType']: ls['WeightedCapacity'] for ls in launch_specifications}
+
+
+def filter_sfr_slaves(sorted_slaves, sfr):
+    sfr_ips = get_sfr_instance_ips(sfr)
+    log.debug("IPs in SFR: {0}".format(sfr_ips))
+    sfr_sorted_slaves = [slave for slave in sorted_slaves if slave_pid_to_ip(slave['pid']) in sfr_ips]
+    sfr_sorted_slave_instances = []
+    for slave in sfr_sorted_slaves:
+        ip = slave_pid_to_ip(slave['pid'])
+        sfr_sorted_slave_instances.append({'ip': ip,
+                                           'pid': slave['pid'],
+                                           'instance_id': get_instance_id_from_ip(ip)})
+    ret = []
+    instance_ids = [slave['instance_id'] for slave in sfr_sorted_slave_instances]
+    instance_descriptions = describe_instances(instance_ids)
+    instance_type_weights = get_instance_type_weights(sfr)
+    for slave in sfr_sorted_slave_instances:
+        instance_description = [instance_description for instance_description in instance_descriptions
+                                if instance_description['InstanceId'] == slave['instance_id']][0]
+        slave['instance_type'] = instance_description['InstanceType']
+        slave['instance_weight'] = instance_type_weights[slave['instance_type']]
+        ret.append(slave)
+    return ret
+
+
+def set_spot_fleet_request_capacity(sfr_id, capacity, dry_run):
+    """ AWS won't modify a request that is already modifying. This
+    function ensures we wait a few seconds in case we've just modified
+    a SFR"""
+    ec2_client = boto3.client('ec2')
+    with Timeout(seconds=AWS_SPOT_MODIFY_TIMEOUT):
+        try:
+            state = None
+            while True:
+                state = get_sfr(sfr_id)['SpotFleetRequestState']
+                if state == 'active':
+                    break
+                log.debug("SFR {0} in state {1}, waiting for state: active".format(sfr_id, state))
+                log.debug("Sleep 5 seconds")
+                time.sleep(5)
+        except TimeoutError:
+            log.error("Spot fleet {0} not in active state so we can't modify it.".format(sfr_id))
+            return
+    if dry_run:
+        return
+    ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=capacity,
```

Comment:
can you return the response object that this returns, so the caller can do something if it fails?